### ISSUE TITLE
kubectl namespace autocompletion not working

### ISSUE THREAD

**Issue by defenestration** [ohmyzsh/ohmyzsh#10581]
### Describe the bug

Hello,

After a recent update to ohmyzsh, i notice this command's autocomplete no longer seems to work. 

### Steps to reproduce

* enable `kubectl` plugin in .zshrc
* use this alias or command, expect to attempt to autocomplete namespaces, however just getting local file list currently.
`kcn`
which is aliased to  
`kubectl config set-context --current --namespace'`

### Expected behavior

Be able to auto-complete kubernetes namespaces. Currently just get local files

### OS / Linux distribution

osx 11.6.2

### Zsh version

5.8

### Oh My Zsh version

master (9c84c344)

### Terminal emulator

iTerm2

### Additional context

This worked before my latest update to OhMyZsh, unsure which version i was using before though.

**Comment by astefanutti** [comment]
I face the same issue, it seems autocompletion is broken in master. It works when I remove the `~/.oh-my-zsh/cache/completions/_kubectl` file, and reload, but only once.

**Comment by astefanutti** [comment]
Resetting the `~/.oh-my-zsh` directory prior to 7ae4f76f6dda1521505c57880ea1e5ee2f1aa183 makes it working back.

**Comment by mcornella** [comment]
That's interesting. In the current version, does it work if you instead source the file?
```sh
source "$ZSH_CACHE_DIR/completions/_kubectl"
```

**Comment by mcornella** [comment]
My `minikube` install does not work so I cannot test this. Please be aware that I'm can't test a solution without your input.

**Comment by defenestration** [comment]
I don't get any errors sourcing the file `source "$ZSH_CACHE_DIR/completions/_kubectl"` but the namespace completions for this command still do not work.

I should note other autocompletions i have tried do work still.  a few i tried just now. 
```
kubectl get
kubectl get pods
kubectl get ns
```

**Comment by Ka0o0** [comment]
> That's interesting. In the current version, does it work if you instead source the file?
> 
> ```shell
> source "$ZSH_CACHE_DIR/completions/_kubectl"
> ```

Hi, auto completion works for me again after sourcing this.

**Comment by zorgzerg** [comment]
Helm plugin broken too. 

Workaround:
```bash
source "$ZSH_CACHE_DIR/completions/_helm"
```

**Comment by dberuben** [comment]
Same issue since i updated from the latest version.
My kubectl autocompletion doesn't work and my iterm2 windows is crashing. I revert back to `d3bb52d7` and it works

**Comment by mcornella** [comment]
Can you test if #10594 works? Try with `omz pr test 10594`.

**Comment by Ka0o0** [comment]
@mcornella yes, this works for me

**Comment by defenestration** [comment]
> Can you test if #10594 works? Try with `omz pr test 10594`.

I'm afraid this doesn't work for me. Tried setting my .zshrc to the default template and added kubectl plugin. But the `kcn` alias does not seem to autocomplete still.

**Comment by mcornella** [comment]
Does it work if you complete namesapce like this?
```
kubectl --namespace <TAB> config set-context --current
```

**Comment by defenestration** [comment]
No, i am just getting local file names as options for tab completing. But i'm wondering if this is an upstream issue perhaps. Looking at https://github.com/kubernetes/kubernetes/pull/106739 

I am using kubectl 1.22 from brew. I checked with 1.23 also from brew and used `source <(kubectl completion zsh)` but still lack the completion it seems. So maybe this is not a zsh issue since other completions work fine. 

Testing in a docker container and some old kubectl versions and loading completions with `source <(kubectl completion zsh)`  it seems the completion worked in kubectl 1.21. If the zsh completion just gets the latest kubectl completion then I think that explains the issue, since 1.23 is the latest.  I think the above issue they may fix it in kubectl 1.24.

**Comment by mcornella** [comment]
That looks like it, yeah. Thanks for researching this and posting an answer!

**Comment by jin-harmoney** [comment]
Upgrading kubectl to 1.24 didn't solve this for me.

Works:

* `kgp TAB`
* `kgp --namespace TAB`
* `kubectl --namespace <TAB> config set-context --current` as @mcornella suggested

Doesn't work:
* `kcn TAB`

**Comment by rparree** [comment]
I am experiencing the same problem

- The update to kubectl v1.24.3 did not work (i did remove `rm -fr cache/completions/_kubectl` before) 
- As others, i have the same problem with helm.
- manually sourcing indeed works (`source "$ZSH_CACHE_DIR/completions/_kubectl"`)
- reverting  `kubectl.plugin.zsh`  prior to 7ae4f76f as indicated by @astefanutti  does to work either (https://raw.githubusercontent.com/ohmyzsh/ohmyzsh/d3bb52d7d825f2a6ce2e1c76ca472b05c6f27b40/plugins/kubectl/kubectl.plugin.zsh)
- this is on the current version 3668ec2a (therefore includes https://github.com/ohmyzsh/ohmyzsh/pull/10594)

Any thoughts?

**Comment by dben** [comment]
Just as a note, After I installed rancher desktop this happened to me. I found out I needed to move 
```bash
### MANAGED BY RANCHER DESKTOP START (DO NOT EDIT)
export PATH="/Users/davidbenson/.rd/bin:$PATH"
### MANAGED BY RANCHER DESKTOP END (DO NOT EDIT)
```
from the end of my .zshrc to before the line for sourcing oh-my-zsh:
```bash
source $ZSH/oh-my-zsh.sh
```

**Comment by h0jeZvgoxFepBQ2C** [comment]
I still have this problem..

```kubectl --namespace <TAB> config set-context --current``` shows the namespace list,

butn ```kcn <TAB>``` shows only the local files list?

**Comment by mfranzil** [comment]
Same for me, both on macOS, Ubuntu 20.04 and 22.04. Issue should be reopened?

**Comment by michal-tomczak** [comment]
Same issue for me on WSL2 Ubuntu. Can we reopen the issue?

**Comment by mcornella** [comment]
Unfortunately we are following the command structure as documented in [1] (also in cheatsheet [2]). The fact that it is not working is an upstream bug that should be reported in [the official repository](https://github.com/kubernetes/kubectl).

It is also not possible for us to change the alias into a form that allows completion of namespaces, as moving the `--namespace` flag would require using a function and completion would not work either.

[1]: https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#-em-set-context-em-

[2]: https://kubernetes.io/docs/reference/kubectl/cheatsheet/#kubectl-context-and-configuration

**Comment by chrko** [comment]
It will be fixed in the upcoming release v1.31 of Kubernetes: https://github.com/kubernetes/kubernetes/pull/124994 🥳🎉
